utils/include/utils/text_matchers.h

```diff
@@ -0,0 +1,257 @@
+    explicit constexpr basic_string_matcher(std::string_view to_match)
+        : to_match(to_match)
+    {}
+
+    template<typename It>
+    bool operator()(It& b, const It& e) const noexcept
+    {
+        It work = b;
+
+        for (auto c : to_match)
+        {
+            if (work == e)
+                return false;
+            if constexpr (single_line)
+            {
+                if (!same_line(work, b))
+                    return false;
+            }
+            auto in_c = *work++;
+            if constexpr (!case_sensitive)
+                in_c = std::toupper((unsigned char)in_c);
+
+            if (in_c != c)
+                return false;
+        }
+        b = work;
+        return true;
+    }
+};
+
+template<bool negate>
+class char_matcher_impl
+{
+    std::string_view to_match;
+
+public:
+    // case_sensitive => to_match assumed in capitals
+    explicit constexpr char_matcher_impl(std::string_view to_match)
+        : to_match(to_match)
+    {}
+
+    template<typename It>
+    constexpr bool operator()(It& b, const It& e) const noexcept
+    {
+        if (b == e)
+            return false;
+        bool result = (to_match.find(*b) != std::string_view::npos) != negate;
+        if (result)
+            ++b;
+        return result;
+    }
+};
+
+using char_matcher = char_matcher_impl<false>;
+using not_char_matcher = char_matcher_impl<true>;
+
+template<typename Matcher>
+constexpr auto star(Matcher&& matcher)
+{
+    return [matcher = std::forward<Matcher>(matcher)]<typename It>(It& b, const It& e) noexcept {
+        while (matcher(b, e))
+            ;
+        return true;
+    };
+}
+
+template<typename Matcher>
+constexpr auto plus(Matcher&& matcher)
+{
+    return [matcher = std::forward<Matcher>(matcher)]<typename It>(It& b, const It& e) noexcept {
+        bool matched = false;
+        while (matcher(b, e))
+            matched = true;
+
+        return matched;
+    };
+}
+
+template<size_t min, size_t max = min, typename Matcher>
+constexpr auto times(Matcher&& matcher) requires(min <= max)
+{
+    return [matcher = std::forward<Matcher>(matcher)]<typename It>(It& b, const It& e) noexcept {
+        auto work = b;
+        size_t count = 0;
+        while (count < max && matcher(work, e))
+            ++count;
+        if (count < min)
+            return false;
+        b = work;
+        return true;
+    };
+}
+
+template<bool empty_allowed, bool single_line>
+class space_matcher
+{
+public:
+    template<typename It>
+    constexpr bool operator()(It& b, const It& e) const noexcept
+    {
+        auto work = b;
+        while (work != e)
+        {
+            if constexpr (single_line)
+            {
+                if (!same_line(work, b))
+                    break;
+            }
+            if (*work != ' ')
+                break;
+            ++work;
+        }
+        if constexpr (!empty_allowed)
+        {
+            if (work == b)
```

Comment:
This checks if the work iterator has not moved from the initial position => failure if empty is not allowed

Reply:
Oh, I thought that it was supposed to check if there is still something left after the spaces were consumed...

Maybe consider renaming `empty_allowed` to `no_match_allowed` or similar

Reply:
I think I originally even called it `no_match_allowed` but then I wanted to avoid the double negative in the code... maybe if I switch the words to `allow_empty` then the code would read `(not allow_empty)`?

Reply:
I guess it would be a bit better but I still have a mental image of an empty line.

What about `allow_empty_match` or `empty_match_allowed`?

Or maybe just a comment would be enough?